retrieval/scrape.go

```diff
@@ -460,18 +472,22 @@ func (sl *scrapeLoop) run(interval, timeout time.Duration, errc chan<- error) {
 
 	buf := bytes.NewBuffer(make([]byte, 0, 16000))
 
+mainLoop:
 	for {
 		buf.Reset()
 		select {
 		case <-sl.ctx.Done():
+			close(sl.stopped)
 			return
+		case <-sl.scrapeCtx.Done():
+			break mainLoop
 		default:
 		}
 
 		var (
 			total, added int
 			start        = time.Now()
-			scrapeCtx, _ = context.WithTimeout(sl.ctx, timeout)
+			scrapeCtx, _ = context.WithTimeout(sl.scrapeCtx, timeout)
```

Comment:
This made me notice that the fix for context cancellation in https://github.com/prometheus/prometheus/pull/2599/files is still out for review...